cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx

```diff
@@ -0,0 +1,145 @@
+        });
+
+        this.initialSetup();
+        this.updateCanvas();
+        this.animateCanvas();
+    }
+
+    public componentDidUpdate(prevProps: Props): void {
+        const { frameData, annotations, curZLayer } = this.props;
+
+        if (
+            prevProps.annotations !== annotations ||
+            prevProps.frameData !== frameData ||
+            prevProps.curZLayer !== curZLayer
+        ) {
+            this.updateCanvas();
+        }
+    }
+
+    public componentWillUnmount(): void {
+        window.removeEventListener('resize', this.fitCanvas);
+        cancelAnimationFrame(this.animateID);
+    }
+
+    private fitCanvas = (): void => {
+        const { canvasInstance } = this.props;
+        canvasInstance.fitCanvas();
+    };
+
+    private onCanvasSetup = (): void => {
+        const { onSetupCanvas } = this.props;
+        onSetupCanvas();
+    };
+
+    public animateCanvas(): void {
+        const { canvasInstance } = this.props;
+
+        canvasInstance.render();
+        this.animateID = requestAnimationFrame(this.animateCanvas.bind(this));
+    }
+
+    private updateCanvas(): void {
+        const {
+            curZLayer, annotations, frameData, canvasInstance,
+        } = this.props;
+
+        if (frameData !== null) {
+            canvasInstance.setup(
+                frameData,
+                annotations.filter((e) => e.objectType !== ObjectType.TAG),
+                curZLayer,
+            );
+        }
+    }
+
+    private initialSetup(): void {
+        const { canvasInstance } = this.props;
+
+        // Size
+        window.addEventListener('resize', this.fitCanvas);
+        this.fitCanvas();
+
+        // Events
+        canvasInstance.html().addEventListener('canvas.setup', () => {}, { once: true });
+        canvasInstance.html().addEventListener('canvas.setup', this.onCanvasSetup);
+    }
+
+    public render(): JSX.Element {
+        const {
+            frame, contextImageHide, getContextImage, loaded, data,
+        } = this.props;
+
+        return (
+            <Layout.Content style={{ position: 'relative' }}>
+                <GlobalHotKeys />
+                {/*
+                    This element doesn't have any props
+                    So, React isn't going to rerender it
+                    And it's a reason why cvat-canvas appended in mount function works
+                */}
+                <ContextImage
+                    frame={frame}
+                    contextImageHide={contextImageHide}
+                    getContextImage={getContextImage}
+                    loaded={loaded}
+                    data={data}
+                />
+                <div
+                    className='cvat-canvas-container'
+                    style={{
+                        overflow: 'hidden',
+                        width: '100%',
+                        height: '100%',
+                    }}
```

Comment:
**style** doesn't contains any dynamic variables, so, please use the css class in places like this